R/conditions.R

```diff
@@ -0,0 +1,815 @@
+#'
+#' fn <- function(valid_input) {
+#'   if (valid_input) {
+#'     return("normal value")
+#'   }
+#'
+#'   # We decide to return the empty string "" as default value. An
+#'   # altenative strategy would be to signal an error. In any case,
+#'   # we want to provide a way for the caller to get a different
+#'   # output. For this purpose, we provide two restart functions that
+#'   # returns alternative defaults:
+#'   restarts <- list(
+#'     rst_empty_chr = function() character(0),
+#'     rst_null = function() NULL
+#'   )
+#'
+#'   with_restarts(.restarts = restarts, {
+#'
+#'     # Signal a typed condition to let the caller know that we are
+#'     # about to return an empty string as default value:
+#'     cnd_signal("default_empty_string")
+#'
+#'     # If no jump to with_restarts, return default value:
+#'     ""
+#'   })
+#' }
+#'
+#' # Normal value for valid input:
+#' fn(TRUE)
+#'
+#' # Default value for bad input:
+#' fn(FALSE)
+#'
+#' # Change the default value if you need an empty character vector by
+#' # defining an inplace handler that jumps to the restart. It has to be
+#' # inplace because thrown handlers jump to the place where they are
+#' # established before being executed, and the restart is not defined
+#' # anymore at that point:
+#' rst_handler <- inplace(function(c) rst_jump("rst_empty_chr"))
+#' with_handlers(fn(FALSE), default_empty_string = rst_handler)
+#'
+#' # You can use restarting() to create restarting handlers easily:
+#' with_handlers(fn(FALSE), default_empty_string = restarting("rst_null"))
+with_restarts <- function(.expr, ..., .restarts = list()) {
+  restarts <- c(list(...), .restarts)
+  with_restarts_(f_capture(.expr), restarts)
+}
+#' @rdname with_restarts
+#' @export
+with_restarts_ <- function(.expr, .restarts = list(), .env = NULL) {
+  f <- as_quoted_f(.expr, .env)
+  f <- env_set(f_interp(~withRestarts(uq(f), uqs(.restarts))), f)
+  f_eval(f)
+}
+
+
+# handlers -----------------------------------------------------------
+
+#' Create a thrown or in place handler.
+#'
+#' There are two types of condition handlers: exiting handlers, which
+#' are thrown to the place where they have been established (e.g.,
+#' \code{\link{with_handlers}()}'s evaluation frame), and local
+#' handlers, which are executed in place (e.g., where the condition
+#' has been signalled). \code{thrown()} and \code{inplace()} create
+#' handlers suitable for \code{\link{with_handlers}()}.
+#'
+#' A subtle point in the R language is that conditions are not thrown,
+#' handlers are. \code{\link[base]{tryCatch}()} and
+#' \code{\link{with_handlers}()} actually catch handlers rather than
+#' conditions. When a critical condition signalled with
+#' \code{\link[base]{stop}()} or \code{\link{panic}()}, R inspects the
+#' handler stack and looks for a handler that can deal with the
+#' condition. If it finds a throwable handler, it throws it to the
+#' function that established it (\code{\link{with_handlers}()}). If R
```

Comment:
I'm not sure anymore that throw is the right term. I was trying to convey that when R finds an exiting handler, it jumps to the stack frame where the handler was defined and calls the handler there. The stack is unwound before calling thrown handlers, similarly to exception systems in other languages.

Other ideas:

- `exiting(function(c) NULL)`? Or `jumping()`. But exiting/jumping where? The throw analogy has the advantage that it implies a target.

- `unwinding(function(c) NULL)`? Probably too technical.

- `rewinding(function(c) NULL)` might be more suggestive than unwinding (to people that had tapes when they were young anyway).


Maybe throw is best, as the distinction between thrown conditions and thrown handlers is a good way of describing how R is different from other languages. It may need more explanation that it really means a jump.